src/components/icon/icon-registry.ts

```diff
@@ -0,0 +1,239 @@
+  */
+class IconConfig {
+  svgElement: SVGElement = null;
+  constructor(public url: string) {
+  }
+}
+
+@Injectable()
+export class MdIconRegistry {
+  // IconConfig objects and cached SVG elements for individual icons.
+  // First level of cache is namespace (which is the empty string if not specified).
+  // Second level is the icon name within the namespace.
+  private _iconConfigs = new Map<string, Map<string, IconConfig>>();
+
+  // IconConfig objects and cached SVG elements for icon sets.
+  // Multiple icon sets can be registered under the same namespace.
+  private _iconSetConfigs = new Map<string, IconConfig[]>();
+
+  // Cache for icons loaded by direct URLs.
+  private _cachedIconsByUrl = new Map<string, SVGElement>();
+
+  private _fontClassNamesByAlias = new Map<string, string>();
+
+  private _inProgressUrlFetches = new Map<string, Observable<string>>();
+
+  private _defaultFontSetClass = 'material-icons';
+
+  constructor(private _http: Http) {
+  }
+
+  public addIcon(iconName: string, url: string): this {
+    return this.addIconInNamespace('', iconName, url);
+  }
+
+  public addIconInNamespace(namespace: string, iconName: string, url: string): this {
+    let iconSetMap = this._iconConfigs.get(namespace);
+    if (!iconSetMap) {
+      iconSetMap = new Map<string, IconConfig>();
+      this._iconConfigs.set(namespace, iconSetMap);
+    }
+    iconSetMap.set(iconName, new IconConfig(url));
+    return this;
+  }
+
+  public addIconSet(url: string): this {
+      return this.addIconSetInNamespace('', url);
+  }
+
+  public addIconSetInNamespace(namespace: string, url: string): this {
+    const config = new IconConfig(url);
+    if (this._iconSetConfigs.has(namespace)) {
+      this._iconSetConfigs.get(namespace).push(config);
+    } else {
+      this._iconSetConfigs.set(namespace, [config]);
+    }
+    return this;
+  }
+
+  public registerFontSet(alias: string, className?: string): this {
+    this._fontClassNamesByAlias.set(alias, className || alias);
+    return this;
+  }
+
+  public setDefaultFontSetClass(className: string) {
+    this._defaultFontSetClass = className;
+    return this;
+  }
+
+  public getDefaultFontSetClass(): string {
+    return this._defaultFontSetClass;
+  }
+
+  loadIconFromNamespaceByName(namespace: string, iconName: string): Observable<SVGElement> {
+    // Return (copy of) cached icon if possible.
+    if (this._iconConfigs.has(namespace) && this._iconConfigs.get(namespace).has(iconName)) {
+      const config = this._iconConfigs.get(namespace).get(iconName);
+      if (config.svgElement) {
+        // We already have the SVG element for this icon, return a copy.
+        return Observable.of(config.svgElement.cloneNode(true));
+      } else {
+        // Fetch the icon from the config's URL, cache it, and return a copy.
+        return this._loadIconFromConfig(config)
+          .do((svg: SVGElement) => config.svgElement = svg)
```

Comment:
The type `SVGElement` shouldn't be necessary here since TS will infer it from the generic, no?